notebooks/kaggle/store-sales-time-series-forecasting/experiments/linear_regression_v1.ipynb

```diff
@@ -0,0 +1,514 @@
+    "import pandas as pd\n",
+    "import matplotlib.pyplot as plt\n",
+    "import matplotlib.dates as mdates\n",
+    "import seaborn as sns\n",
+    "import sklearn\n",
+    "import pdpipe as pdp\n",
+    "from pdpipe import df\n",
+    "from pathlib import Path\n",
+    "from sklearn.linear_model import LinearRegression, Ridge, Lasso\n",
+    "from sklearn.preprocessing import OneHotEncoder, MinMaxScaler\n",
+    "from sklearn.pipeline import make_pipeline\n",
+    "from sklearn.model_selection import cross_validate, cross_val_score, TimeSeriesSplit, GridSearchCV\n",
+    "from sklearn.compose import ColumnTransformer\n",
+    "from sklearn.metrics import mean_squared_log_error, mean_squared_error, mean_absolute_error, r2_score, mean_absolute_percentage_error\n",
+    "\n",
+    "%matplotlib inline\n",
+    "warnings.simplefilter(action='ignore', category=FutureWarning)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "94082fbb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "DATA_ROOT = Path('data', 'kaggle', 'store-sales-time-series-forecasting')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6a75808e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train = pd.read_csv(DATA_ROOT / 'prepared_train.csv', low_memory=False)\n",
+    "train.head()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b2f263f7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "X = train[['store_nbr', 'family', 'onpromotion', 'dcoilwtico']]\n",
+    "y = train['sales']\n",
+    "print(X.shape, y.shape)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "904053a8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "categorical_columns = ['store_nbr', 'family']\n",
+    "numerical_columns = X.columns.drop(categorical_columns)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6f5368cc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "transform_pipeline = pdp.PdPipeline([\n",
+    "        pdp.Scale('MinMaxScaler', numerical_columns),\n",
+    "        pdp.OneHotEncode(categorical_columns),\n",
+    "    ])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1cc5a8b3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "X = transform_pipeline.apply(X)\n",
+    "X.shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "74aec17b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ts_cv = TimeSeriesSplit(\n",
+    "    n_splits=4,\n",
+    "    gap=0,\n",
+    "    max_train_size=365*54*33,\n",
+    "    test_size=15*54*33,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f3652706",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def cv_algorithm(X=X, model_cv=LinearRegression()):\n",
```

Comment:
I wrote this function, because I needed to set all the negative predictions to zero. I have no idea how to do it with the help of these sklearn utilities

Reply:
Ok, understand. I suggested you an option of implementing a custom meta-regressor (see the comment below), hope this will help.